Fix startup crash with --io-top and ignore duplicate plugin filenames

Lekensteyn opened the pull request:
##### DOOL VERSION
dool v1.3.4

##### SUMMARY
Fixes a regression from v1.3.3 where `dool --top-io` crashes on startup:
```
$ dool --top-io
Traceback (most recent call last):
  File "/usr/bin/dool", line 3088, in <module>
    __main()
    ~~~~~~^^
  File "/usr/bin/dool", line 3076, in __main
    main()
    ~~~~^^
  File "/usr/bin/dool", line 2784, in main
    scheduler.run()
    ~~~~~~~~~~~~~^^
  File "/usr/lib/python3.13/sched.py", line 151, in run
    action(*argument, **kwargs)
    ~~~~~~^^^^^^^^^^^^^^^^^^^^^
  File "/usr/bin/dool", line 2890, in perform
    o.extract()
    ~~~~~~~~~^^
  File "<string>", line 53, in extract
KeyError: 'rchar:'
```

While at it, I also ignored duplicate plugin filenames. That makes it possible to test this fix without having to remove the dool package installed through the package manager on Arch Linux.

In case you are wondering, the `dool --list` code does not have to be modified since it compares the file location of the loaded plugin against the directory from the `pluginpath` directory.

Lekensteyn committed:
Fix crash when using --top-io plugin

Use the same field names during reset, otherwise dool crashes with a
KeyError due to a missing `rchar:` dict key.

Fixes: 5068abef0a96 ("Clarify what io/bio mean and document where to get more info")

Lekensteyn committed:
Ensure duplicate plugin files are ignored

When I run `./dool --top-io` from the Git repository, I expect a plugin
to be loaded from the `plugins/` subdirectory, not `/usr/share/dool/`
which contains an older version installed through the package manager.
Prioritize `/usr/local` as that is likely newer than `/usr`.

scottchiefbaker commented:
This is really interesting, we just updated this in 5068abe. How is it broken already? @emandret didn't we test this?

emandret commented:
Interesting, maybe some systems don't have rchar and wchar. What system is OP using?

Lekensteyn commented:
> This is really interesting, we just updated this in [5068abe](https://github.com/scottchiefbaker/dool/commit/5068abef0a960c173a91e84a45e7fea1b0a802c0). How is it broken already? @emandret didn't we test this?

I guess it was only tested with `--top-bio` where it uses `read_bytes`. In the linked commit (also mentioned in my commit message), the same code change was copied from `dool_top_bio.py` to `dool_top_io.py`. I didn't check whether `rchar` or `read_bytes` would be more appropriate, but I just fixed the inconsistency that caused a crash.

> Interesting, maybe some systems don't have rchar and wchar. What system is OP using?

I do have the wchar and rchar fields, running Arch Linux with linux 6.13.0-arch1-1. Initially the process list is empty, so `pidlist1` will be empty. Then `pidset2` is populated with current processes. Finally, reading `self.pidset1[pid]['rchar:']` will fail as there is no data. Did you try running `dool --top-io`, did it work for you?

emandret commented:
> I do have the wchar and rchar fields, running Arch Linux with linux 6.13.0-arch1-1. Initially the process list is empty, so `pidlist1` will be empty. Then `pidset2` is populated with current processes. Finally, reading `self.pidset1[pid]['rchar:']` will fail as there is no data. Did you try running `dool --top-io`, did it work for you?

Yes it did, I am running a Fedora virtual machine. I think we need some guards to avoid accessing keys that may not exist in an empty dictionary. Oddly, that behaviour didn't happen on my environment, and I wrongly assumed it would port to other Linux distros with no issues.

Lekensteyn commented:
I still managed to reproduce it on a Fedora 41 (Cloud Edition) VM with Python 3.13.0 and Linux 6.11.4-301.fc41.x86_64, both as regular and as root user.

Regular users cannot open `/proc/1/io`, apparently `linecache.getline("/proc/1/io", 1)` (via `proc_splitlines("/proc/1/io")`) returns an empty string in that case. Therefore `self.pidset2` will also not be populated with `rchar:` and `wchar:`. The plugin code would continue if an `IOError` is raised, but that currently does not happen:
```
$ python3 -c 'import platform, linecache; print("%r %s" % (linecache.getline("/proc/self/io", 1), platform.python_version()))'
'rchar: 917095\n' 3.13.2
$ python3 -c 'import platform, linecache; print("%r %s" % (linecache.getline("/proc/1/io", 1), platform.python_version()))'
'' 3.13.2
$ python2 -c 'import platform, linecache; print("%r %s" % (linecache.getline("/proc/1/io", 1), platform.python_version()))'
'' 2.7.18
```

Do you see other PIDs in `/proc/`? If not, have you set the `hidepid=4` mount option on `/proc` that would hide any `/proc/<pid>` directory that the user cannot ptrace.

Are you sure that you are running the patched dool plugin? Before this PR, it would prefer `/usr/local/dool/dool_top_io.py` and `/usr/share/dool/dool_top_io.py` over whatever is installed in Git.

Based on static code analysis, I don't think that `dool --top-io` in dool v1.3.4 would ever have worked, correct me if I am wrong?
```python
    def extract(self):
        self.output       = ''
        # ASSUMPTION 1: pidset1 is initialized in __init__, both are not touched except in this function. 
        self.pidset2      = {}
        self.val['usage'] = 0.0
        # ASSUMPTION 2: `proc_pidlist` returns some PIDs from /proc except for the dool process itself
        for pid in proc_pidlist():
            try:
                ### Reset values
                if pid not in self.pidset2:
                    self.pidset2[pid] = {'read_bytes:': 0, 'write_bytes:': 0}
                if pid not in self.pidset1:
                    self.pidset1[pid] = {'read_bytes:': 0, 'write_bytes:': 0}
                # ASSUMPTION 3: initially both pidset2[pid] and pidset1[pid] = {'read_bytes:': 0, 'write_bytes:': 0}

                ### Extract name
                name = proc_splitline('/proc/%s/stat' % pid)[1][1:-1]

                ### Extract counters
                # ANALYSIS: proc_splitlines returns an empty list for /proc/1/io as non-root user
                for l in proc_splitlines('/proc/%s/io' % pid):
                    if len(l) != 2: continue
                    self.pidset2[pid][l[0]] = int(l[1])
                # ASSUMPTION 4: pidset2[pid] = {'read_bytes:': 0, 'write_bytes:': 0, 'rchar:' 123, 'wchar:': 456}
                #   or unchanged if pid == "1" and running as non-root user
            except IOError:
                # ANALYSIS: seems unreachable through proc_splitline and proc_splitlines
                continue
            except IndexError:
                # ANALYSIS: seems unreachable assuming that /proc/<pid>/stat does not contain garbage
                continue

            if (op.bits):
                factor = 8
            else:
                factor = 1

            # INFO: https://www.kernel.org/doc/html/latest/filesystems/proc.html#proc-pid-io-display-the-io-accounting-fields
            #
            ### 'rchar' counts bytes read from the task POV, e.g. open files which may be read from page cache, reading from a socket or pipe
            # ANALYSIS: self.pidset1[pid]['rchar:'] is always unset based on assumption 3
            # ANALYSIS: self.pidset2[pid]['rchar:'] may be unset based on assumption 4
            # ANALYSIS: therefore a KeyError is expected
            read_usage  = (self.pidset2[pid]['rchar:'] - self.pidset1[pid]['rchar:']) * factor / elapsed
```

scottchiefbaker commented:
I think the issue here is that we're trying to read `/proc/{num}/io` on low numbered PIDs which are owned by root and we don't have access as a regular user. I think we just need to be smarter about erroring out if we don't get anything back from reading `/proc/`

scottchiefbaker commented:
I just landed some potential fixes in [db33037](https://github.com/scottchiefbaker/dool/commit/db330372733b9f79a67acdcea5219576356281dc) can someone test please?

scottchiefbaker commented:
With these fixes I'm seeing consistency across our IO plugins now:

```text
bakers@snikt(~/github/dool)
$ ./dool --top-io --top-bio --top-io-adv 1 5
┄┄┄┄┄┄┄┄top-io┄┄┄┄┄┄┄┄┬┄┄┄┄┄┄top┄block-io┄┄┄┄┄┬┄┄┄┄┄┄┄┄┄most┄expensive┄i/o┄process┄┄┄┄┄┄┄┄┄
name        read write│name         read write│process                 pid  read write  cpu
                      │firefox      285B 8230B│firefox               83169  175K   16K 0.1%
firefox    8723K   40B│                       │firefox               94751 1090K    5B   0%
firefox     128B 2533K│firefox        0   320K│firefox               83169   16B  317K 0.1%
firefox      77M   96K│firefox        0    20K│firefox               83169 9819K   12K 4.9%
firefox    1540K  389K│firefox        0    48K│firefox               83169  193K   49K 6.5%
```

scottchiefbaker commented:
I'm going to close this issue as fixed. Please open a separate issue and we discuss duplicate plugins

Lekensteyn commented:
> I think the issue here is that we're trying to read `/proc/{num}/io` on low numbered PIDs which are owned by root and we don't have access as a regular user. I think we just need to be smarter about erroring out if we don't get anything back from reading `/proc/`

Did you read my analysis above? While the fix in commit db330372733b9f79a67acdcea5219576356281dc fixes the crash, it introduces a new bug where the results from the first read are discarded. Hence the empty column in your first row. I can submit a fix for that too, it will be the same as this PR.

For the plugin issue, is it okay to open a PR right away or do you prefer an issue tracking the problem?

scottchiefbaker commented:
Good call on the first row thing... I was more focused on it not crashing. I'll see about fixing that today (it's easy).

Let's start with an issue for the plugins. It's part of a larger discussion I think, so let's get input on it.

scottchiefbaker commented:
@Lekensteyn will you submit a fix for *just* the missing first column? I'll land that asap. Make sure the PR targets the `next` branch.

Lekensteyn commented:
> @Lekensteyn will you submit a fix for _just_ the missing first column? I'll land that asap. Make sure the PR targets the `next` branch.

Sure thing, I will prepare something in a bit!

And I will also create an issue for the plugin discussion.

scottchiefbaker commented:
Excellent. I appreciate your enthusiasm to the project. You got me reinvigorated. I landed 20 little commits on `dool` yesterday

Lekensteyn commented:
Haha, it is a useful tool, thanks for taking the initiative in maintaining it!

I will try to help where I can, feel free to poke me through various channels if I don't reply in time.
